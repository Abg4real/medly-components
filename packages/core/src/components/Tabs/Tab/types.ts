import { HTMLProps } from '@medly-components/utils';
import { TabSize } from '../types';

export interface StyledProps extends HTMLProps<HTMLButtonElement> {
    active?: boolean;
    hasIcon?: boolean;
    tabSize?: TabSize;
}

export interface Props extends StyledProps {
    /** Id of tab */
    id: any;
    /** Label of the tab */
    label: string;
    /** To be used to hide the tab */
    hide?: boolean;
    /** Icon to be shown */
    icon?: React.ElementType;
}
