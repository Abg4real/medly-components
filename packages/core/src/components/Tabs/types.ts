import { HTMLProps, WithStyle } from '@medly-components/utils';
import { Props as TabProps } from './Tab/types';
import { Props as TabListProps } from './TabList/types';
import { Props as TabPanelProps } from './TabPanel/types';

export type TabSize = 'S' | 'M' | 'L';

export type TabBackground = 'WHITE' | 'GREY';

export type TabStyle = 'OPEN' | 'CLOSED';

export interface Props extends HTMLProps<HTMLDivElement> {
    /** Id of the default active tab */
    defaultActive?: any;
    /** Id of the active tab */
    active?: any;
    /**  Function to be called on selecting any tab */
    onChange?: (id: any) => void;
    /** Size for tab, default is S */
    tabSize?: TabSize;
    /** Tab can have tab style of OPEN and CLOSED */
    tabStyle?: TabStyle;
    /**  Tab Background can be WHITE or GREY */
    tabBackground?: TabBackground;
}

export interface StaticProps {
    Tab?: React.FC<TabProps> & WithStyle;
    TabList?: React.FC<TabListProps> & WithStyle;
    TabPanel?: React.FC<TabPanelProps> & WithStyle;
}
