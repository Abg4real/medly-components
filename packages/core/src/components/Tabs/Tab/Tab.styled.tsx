import { SvgIcon } from '@medly-components/icons';
import { Theme } from '@medly-components/theme/src';
import { css, styled } from '@medly-components/utils';
import { TextStyled } from '../../Text/Text.styled';
import { TabSize } from '../types';
import { StyledProps } from './types';

const getTabSpacing = (tabSize: TabSize, hasIcon = false) => {
    const verticalSpacing = 15;
    let spacing = verticalSpacing;

    if (tabSize === 'L') {
        spacing = verticalSpacing + 8;
    }

    if (hasIcon) {
        return spacing - 1;
    }
    return spacing;
};

const activeStyle = (theme: Theme, active: boolean, tabSize: TabSize) => {
    const {
        tabs: { textColor, backgroundColor, backgroundTheme, tabStyle }
    } = theme;

    if (backgroundTheme === 'WHITE' && tabStyle === 'CLOSED') {
        return css``;
    } else {
        return css`
            color: ${textColor.active};
            border-bottom: 4px solid ${textColor.active};
            ${SvgIcon} {
                * {
                    fill: ${active && tabSize !== 'S' ? backgroundColor.nonActive : textColor.active};
                }
            }
            &:hover {
                background-color: ${backgroundColor.nonActive};
                ${TextStyled} {
                    color: ${textColor.active};
                }
            }
        `;
    }
};

const disabledStyle = (theme: Theme) => {
    return css`
        opacity: 0.4;
        &:hover {
            background-color: transparent;
            ${TextStyled} {
                color: ${theme.tabs.textColor.nonActive};
            }
        }
    `;
};

const nonActiveStyle = (theme: Theme) => {
    const {
        tabs: { textColor, backgroundColor }
    } = theme;
    return css`
        color: ${textColor.nonActive};
        &:hover {
            background-color: ${backgroundColor.hovered};
            ${TextStyled} {
                color: ${textColor.hovered};
            }
        }
    `;
};

const svgStyles = (theme: Theme, active: boolean) => {
    const {
        tabs: { textColor }
    } = theme;
    return css`
        ${SvgIcon} {
            border-radius: 50%;
            border: 1px solid ${active ? textColor.active : theme.colors.grey[100]};
            background-color: ${active ? textColor.active : theme.colors.grey[100]};
            padding: 5px;
            font-size: 2rem;
        }
    `;
};

export const Button = styled.button<StyledProps>`
    background-color: transparent;
    border: 0;
    border-bottom: 4px solid transparent;
    padding: ${({ tabSize, hasIcon }) => `${getTabSpacing(tabSize, hasIcon)}px ${16}px`};
    user-select: none;
    text-decoration: none;
    display: flex;
    align-items: center;
    flex-direction: row;
    justify-content: flex-start;

    ${TextStyled} {
        font-size: ${({ tabSize }) => (tabSize === 'S' ? '1.4rem' : '1.6rem')};
        margin-left: ${({ hasIcon }) => (hasIcon ? '16px' : 0)};
    }

    &:hover {
        cursor: pointer;
    }

    &:focus {
        outline: none;
    }

    &:disabled {
        cursor: not-allowed;
    }

    ${({ theme, active, tabSize, hasIcon }) => hasIcon && tabSize !== 'S' && svgStyles(theme, active)}

    ${({ theme, active }) => !active && nonActiveStyle(theme)}

    ${({ theme, active, tabSize }) => active && activeStyle(theme, active, tabSize)}

    ${({ theme, disabled }) => disabled && disabledStyle(theme)}
`;
