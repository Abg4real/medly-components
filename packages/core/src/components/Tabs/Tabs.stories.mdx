import { HTMLProps } from '@medly-components/utils';
import { Tabs } from './Tabs.tsx';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Tabs.stories';

<Meta title="Core/Tabs" component={Tabs} />

# Tabs

### Displays specific content based on selected tab

Use `Tabs.Tab` to add tab in `Tabs` component. Like bellow . . .

```typescript
<Tabs aria-label="Basic Tabs">
    <Tabs.Tab id="tab1" label="Add">
        Content for the add panel
    </Tabs.Tab>
    <Tabs.Tab id="tab2" label="Edit">
        Content for the edit panel
    </Tabs.Tab>
    <Tabs.Tab id="tab3" label="Delete">
        Content for the delete panel
    </Tabs.Tab>
</Tabs>
```

<Preview withToolbar>
    <Story name="Basic">{stories.Basic()}</Story>
</Preview>

### Keyboard Support

It also supports keyboard navigation when it is in focus

| Key         | Function                                      |
| ----------- | --------------------------------------------- |
| Right Arrow | Moves focus to the next tab                   |
| Left Arrow  | Moves focus to the previous tab               |
| Home        | Moves focus to the first tab and activates it |
| End         | Moves focus to the last tab and activates it  |

### WithIcon

We can provide icon also using `icon` prop.

<Preview withToolbar>
    <Story name="WithIcon">{stories.WithIcon()}</Story>
</Preview>

### Uncontrolled

When using uncontrolled `Tabs` component you do not need to provide any prop. By default first tab is active if you haven't provide `defaultActive` prop.

<Preview withToolbar>
    <Story name="Uncontrolled">{stories.Uncontrolled()}</Story>
</Preview>

### Controlled

When using controlled `Tabs` component you have to provide `active` and `onChange` prop.

<Preview withToolbar>
    <Story name="Controlled">{stories.Controlled()}</Story>
</Preview>

### Props

<Props of={Tabs} />

### TabsTheme

<Props of={stories.ThemeInterface} />
