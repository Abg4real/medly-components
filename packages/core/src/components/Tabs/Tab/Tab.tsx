import { WithStyle } from '@medly-components/utils';
import React from 'react';
import Text from '../../Text';
import * as Styled from './Tab.styled';
import { Props } from './types';

export const Tab: React.FC<Props> & WithStyle = React.memo(props => {
    const { id, active, label, icon: Icon = null, ...restProps } = props;

    return (
        <Styled.Button
            {...{ id, active, ...restProps }}
            type="button"
            role="tab"
            aria-selected={active ? 'true' : 'false'}
            aria-controls={`panel-${id}`}
            tabIndex={active ? 0 : -1}
        >
            {Icon && <Icon focusable={false} />}
            <Text>{label}</Text>
        </Styled.Button>
    );
});
Tab.displayName = 'Tab';
Tab.Style = Styled.Button;
