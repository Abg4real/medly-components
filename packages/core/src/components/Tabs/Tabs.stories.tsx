import { AddIcon, DashboardIcon, DateRangeIcon, DeleteIcon, EditIcon, ViewColumnIcon } from '@medly-components/icons';
import { defaultTheme, TabsTheme } from '@medly-components/theme';
import { select } from '@storybook/addon-knobs';
import React, { useState } from 'react';
import { Tabs } from './Tabs';
import { Props } from './types';

export const ThemeInterface: React.FC<TabsTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.tabs
};

const size: Props['tabSize'][] = ['S', 'M', 'L'];

export const Basic = () => (
    <Tabs aria-label="Basic Tabs" tabSize={select('Tab Size', size, 'S')}>
        <Tabs.Tab id="tab1" label="Add" hide>
            Content for the add panel
        </Tabs.Tab>
        <Tabs.Tab id="tab2" label="Edit">
            Content for the edit panel
        </Tabs.Tab>
        <Tabs.Tab id="tab3" label="Delete">
            Content for the delete panel
        </Tabs.Tab>
    </Tabs>
);

export const WithIcon = () => {
    return (
        <Tabs defaultActive="tab1" aria-label="WithIcon Tabs" tabSize={select('Tab Size', size, 'S')}>
            <Tabs.Tab id="tab1" label="Add" icon={AddIcon}>
                Content for the add panel
            </Tabs.Tab>
            <Tabs.Tab id="tab2" label="Edit" icon={EditIcon}>
                Content for the edit panel
            </Tabs.Tab>
            <Tabs.Tab id="tab3" label="Delete" icon={DeleteIcon} disabled={true}>
                Content for the delete panel
            </Tabs.Tab>
        </Tabs>
    );
};

export const WithDescriptionAndCount = () => {
    return (
        <Tabs defaultActive="tab1" aria-label="WithDescriptionAndCount Tabs" tabSize={select('Tab Size', size, 'S')}>
            <Tabs.Tab id="tab1" label="Add" secondaryLabel="Details" count={68} icon={DashboardIcon}>
                Content for the add panel
            </Tabs.Tab>
            <Tabs.Tab id="tab2" label="Edit" secondaryLabel="Details" count={10} icon={DateRangeIcon}>
                Content for the edit panel
            </Tabs.Tab>
            <Tabs.Tab id="tab3" label="Delete" secondaryLabel="Details" count={30} icon={ViewColumnIcon} disabled={true}>
                Content for the delete panel
            </Tabs.Tab>
        </Tabs>
    );
};

export const Uncontrolled = () => {
    return (
        <Tabs defaultActive="tab1" aria-label="Uncontrolled Tabs">
            <Tabs.Tab id="tab1" label="Add">
                Content for the add panel
            </Tabs.Tab>
            <Tabs.Tab id="tab2" label="Edit">
                Content for the edit panel
            </Tabs.Tab>
            <Tabs.Tab id="tab3" label="Delete">
                Content for the delete panel
            </Tabs.Tab>
        </Tabs>
    );
};

export const Controlled = () => {
    const [active, setActiveTab] = useState('tab1');
    return (
        <Tabs active={active} onChange={setActiveTab} aria-label="Controlled Tabs">
            <Tabs.Tab id="tab1" label="Add">
                Content for the add panel
            </Tabs.Tab>
            <Tabs.Tab id="tab2" label="Edit">
                Content for the edit panel
            </Tabs.Tab>
            <Tabs.Tab id="tab3" label="Delete">
                Content for the delete panel
            </Tabs.Tab>
        </Tabs>
    );
};
