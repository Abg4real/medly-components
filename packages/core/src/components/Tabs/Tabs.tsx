import { WithStyle } from '@medly-components/utils';
import React, { useCallback, useMemo, useState } from 'react';
import Tab from './Tab';
import TabList from './TabList';
import TabPanel from './TabPanel';
import * as Styled from './Tabs.styled';
import { Props, StaticProps } from './types';

export const Tabs: React.SFC<Props> & StaticProps & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const { defaultActive, active, onChange, children, ...restProps } = props,
            tabsId = props.id || 'medly-tabs',
            tabIds = useMemo(
                () =>
                    children
                        ? React.Children.toArray(children)
                              .filter((child: any) => !child.props.hide)
                              .map((child: any) => child.props.id)
                        : [undefined],
                [children]
            ),
            [key, setKey] = useState(defaultActive || tabIds[0]),
            activeTab = useMemo(() => active || key, [active, key]);

        const handleTabChange = useCallback(
            (id: any) => {
                !active && setKey(id);
                onChange && onChange(id);
            },
            [active, onChange]
        );

        if (React.Children.count(children) === 0) return null;

        return (
            <Styled.Tabs id={tabsId} ref={ref}>
                <TabList {...restProps} id={`${tabsId}-list`} active={activeTab} onChange={handleTabChange}>
                    {children}
                </TabList>
                <TabPanel id={`${tabsId}-panel`} active={activeTab}>
                    {children}
                </TabPanel>
            </Styled.Tabs>
        );
    })
);
Tabs.displayName = 'Tabs';
Tabs.Style = Styled.Tabs;
Tabs.Tab = Tab;
Tabs.TabList = TabList;
Tabs.TabPanel = TabPanel;
