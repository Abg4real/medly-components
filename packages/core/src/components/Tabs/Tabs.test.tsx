import { fireEvent, render } from '@test-utils';
import React from 'react';
import Tab from './Tab';
import { Tabs } from './Tabs';

describe('Tabs', () => {
    it('should render properly without any prop', () => {
        const { container } = render(
            <Tabs>
                <Tab id="tab1" label="Add">
                    Content for the add panel
                </Tab>
                <Tab id="tab2" label="Edit">
                    Content for the edit panel
                </Tab>
            </Tabs>
        );
        expect(container).toMatchSnapshot();
    });

    it('should not render when there is no children', () => {
        const { container } = render(<Tabs />);
        expect(container).toMatchSnapshot();
    });

    it('should hide the tab if hide prop is given', async () => {
        const { queryByText } = render(
            <Tabs defaultActive="tab2">
                <Tab id="tab1" label="Add" hide>
                    Content for the add panel
                </Tab>
                <Tab id="tab2" label="Edit">
                    Content for the edit panel
                </Tab>
            </Tabs>
        );
        expect(queryByText('Content for the add panel')).toBeNull();
    });

    it('should select expected tab on click on any tab, when only defaultActive prop is given', async () => {
        const { getByText, queryByText } = render(
            <Tabs defaultActive="tab1">
                <Tab id="tab1" label="Add">
                    Content for the add panel
                </Tab>
                <Tab id="tab2" label="Edit">
                    Content for the edit panel
                </Tab>
            </Tabs>
        );
        fireEvent.click(getByText('Edit'));
        expect(queryByText('Content for the edit panel')).toBeInTheDocument();
    });

    it('should call onChange prop with expected tab id', () => {
        const mockOnchange = jest.fn();
        const { getByText } = render(
            <Tabs active="tab1" onChange={mockOnchange}>
                <Tab id="tab1" label="Add">
                    Content for the add panel
                </Tab>
                <Tab id="tab2" label="Edit">
                    Content for the edit panel
                </Tab>
            </Tabs>
        );
        fireEvent.click(getByText('Edit'));
        expect(mockOnchange).toBeCalledWith('tab2');
    });
});
