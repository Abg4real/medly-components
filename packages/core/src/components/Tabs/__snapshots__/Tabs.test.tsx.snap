// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should not render when there is no children 1`] = `<div />`;

exports[`Tabs should render properly without any prop 1`] = `
.c6 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
}

.c1 {
  margin: 0;
  padding: 0;
  border: 0;
  outline: 0;
  background: none;
  color: #005AEE;
  background-color: #ffffff;
  border-right: 0 none;
}

.c1::after {
  content: '';
  width: 100%;
  display: block;
  height: 4px;
  background-color: #005AEE;
  border-radius: 5px 5px 0 0;
}

.c1:hover {
  cursor: pointer;
}

.c1:hover::after {
  background-color: #005AEE;
}

.c1:focus {
  outline: none;
}

.c1:disabled {
  cursor: not-allowed;
}

.c1 .sc-AxirZ * {
  fill: #005AEE;
}

.c1:hover {
  background-color: #ffffff;
}

.c1:hover .c5 {
  color: #005AEE;
}

.c7 {
  margin: 0;
  padding: 0;
  border: 0;
  outline: 0;
  background: none;
  color: #607890;
  border-right: 0 none;
}

.c7::after {
  content: '';
  width: 100%;
  display: block;
  height: 4px;
  background-color: transparent;
  border-radius: 5px 5px 0 0;
}

.c7:hover {
  cursor: pointer;
}

.c7:hover::after {
  background-color: #F7F8F9;
}

.c7:focus {
  outline: none;
}

.c7:disabled {
  cursor: not-allowed;
}

.c7:last-child {
  border-right: 0 none;
}

.c7:hover {
  background-color: #F7F8F9;
}

.c7:hover .c5 {
  color: #435465;
}

.c2 {
  background-color: transparent;
  border: 0;
  padding: 15px 16px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
}

.c2 .c5 {
  font-size: 1.4rem;
  line-height: 2.2rem;
}

.c2 p {
  line-height: 2.2rem;
  font-size: 1.4rem;
}

.c3 {
  margin-left: 0;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 .c5 {
  grid-area: label;
}

.c0 {
  background-color: #ffffff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  overflow-x: auto;
  border: 0 none;
  border-radius: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c8 {
  padding: 10px;
  height: 100%;
}

.c8:focus {
  outline: none;
}

<div>
  <div
    class=""
    id="medly-tabs"
  >
    <div
      class="c0"
      id="medly-tabs-list"
    >
      <button
        aria-controls="panel-tab1"
        aria-selected="true"
        class="c1"
        id="tab1"
        role="tab"
        tabindex="0"
        type="button"
      >
        <div
          class="c2"
        >
          <div
            class="c3"
          >
            <div
              class="c4"
            >
              <span
                class="c5 c6"
              >
                Add
              </span>
            </div>
          </div>
        </div>
      </button>
      <button
        aria-controls="panel-tab2"
        aria-selected="false"
        class="c7"
        id="tab2"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <div
          class="c2"
        >
          <div
            class="c3"
          >
            <div
              class="c4"
            >
              <span
                class="c5 c6"
              >
                Edit
              </span>
            </div>
          </div>
        </div>
      </button>
    </div>
    <div
      aria-labelledby="tab1"
      class="c8"
      id="medly-tabs-panel-tab1"
      role="tabpanel"
      tabindex="0"
    >
      Content for the add panel
    </div>
  </div>
</div>
`;
